docs/proposals/image-promotion.md

```diff
@@ -0,0 +1,201 @@
+of thousand applications.
+
+#### Example
+
+* OpenShift engineering push the `openshift/ruby-22-rhel7` image to internal CI registry
+* OpenShift QA team pull the image from the internal CI registry and run tests
+* OpenShift QA team sign-off the Docker image when tests pass and push it to "integration" registry
+* OpenShift operations team pulls the image from "integration" registry and promote it to "stage" environment
+* OpenShift operations team manually promote the image to production
+
+### Semi-automatic promotion and testing of builder images
+
+As a developer of the S2I image, I want to make sure that the changes I propose won't break
+the applications that are based on this image, when they will be rebuilt.
+For that I want to promote the image as "ok-to-merge", only when the sample application
+that use this image pass the verification checks.
+
+#### Example of pull request tester
+
+* OpenShift engineering propose a change to `openshift/ruby-22-rhel7` image
+* Jenkins Github plugin detects a new pull request and docker build the new testing image
+* Jenkins job connects to a running OpenShift server, that runs the example applications which are based on this image
+* Jenkins job imports the updated image to a running OpenShift server and tag it as "test"
+* A new deployment is triggered for applications that have ImageChangeTrigger set to watch this image
+* As part of the deployment post-lifecycle hook, the applications are verified and the ImageStreamTag is sign-off by the application name
+* Jenkins job waits until all applications sign-off the updated image as "verified"
+* Jenkins Github plugin send notification to original pull request
+
+#### Example of push_images job
+
+* OpenShift engineering merge change to `openshift/ruby-22-rhel7` repository
+* OpenShift server detect the change and execute test suite for given image
+* If the image is 'base' image, OpenShift server launch multiple tests for all dependent images
+* OpenShift server push the image to internal registry and Docker Hub (see: "Manual promotion of builder image")
+
+## Areas of improvement
+
+The first use-case describes a typical Continuos Delivery example, where the developer uses
+both the Jenkins server and the OpenShift to deliver updates to the application.
+
+### Jenkins
+
+* There is no support for enabling the Jenkins when creating a new application
+* There is no official "jenkins" plugin for OpenShift availabl
+* No first class way to give Jenkins credentials needed for `oc` command
+* Users have to manually tweak the Jenkins job parameters
+* The Jenkins job definition is written in shell and it is complex
+
+### Projects
+
+* Having to promote images from between multiple projects (or clusters) has security implication
+
+### Deployments
+
+* The "post" lifecycle hook can be executed only in the image that is being deployed
+* The images must have OpenShift CLI tools installed in order to do tagging
```

Comment:
It would have to in order to do it - but we could make it easier to consume

On Fri, Aug 7, 2015 at 7:24 PM, Ben Parees <notifications@github.com> wrote:

> In docs/proposals/image-promotion.md
> <https://github.com/openshift/origin/pull/4016#discussion_r36569939>:
>
> > +### Jenkins
> > +
> > +* There is no support for enabling the Jenkins when creating a new application
> > +* There is no official "jenkins" plugin for OpenShift availabl
> > +* No first class way to give Jenkins credentials needed for `oc` command
> > +* Users have to manually tweak the Jenkins job parameters
> > +* The Jenkins job definition is written in shell and it is complex
> > +
> > +### Projects
> > +
> > +* Having to promote images from between multiple projects (or clusters) has security implication
> > +
> > +### Deployments
> > +
> > +* The "post" lifecycle hook can be executed only in the image that is being deployed
> > +* The images must have OpenShift CLI tools installed in order to do tagging
>
> can a deploymenthook specify that image rather than one of the container
> images?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4016/files#r36569939>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift